File: edgequake/sdks/ruby/test/e2e_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/edgequake"

# E2E tests for EdgeQuake Ruby SDK.
# Requires running backend at localhost:8080 (or EDGEQUAKE_BASE_URL).
class E2ETest < Minitest::Test
  def setup
    base = ENV.fetch("EDGEQUAKE_BASE_URL", "http://localhost:8080")
    tenant_id = ENV.fetch("EDGEQUAKE_TENANT_ID", "00000000-0000-0000-0000-000000000002")
    user_id = ENV.fetch("EDGEQUAKE_USER_ID", "00000000-0000-0000-0000-000000000001")
    config = EdgeQuake::Config.new(base_url: base, tenant_id: tenant_id, user_id: user_id)
    @client = EdgeQuake::Client.new(config: config)
  end

  # 1. Health
  def test_health_check
    h = @client.health.check
    assert_equal "healthy", h["status"]
    refute_nil h["version"]
  end

  # 2. Documents
  def test_documents_list_and_upload
    list = @client.documents.list
    refute_nil list["documents"]
    refute_nil list["total"]

    resp = @client.documents.upload_text(
      title: "Ruby SDK Test #{SecureRandom.hex(4)}",
      content: "Ruby SDK integration test. Knowledge graphs are powerful."
    )
    refute_nil resp["document_id"]
    refute_nil resp["status"]
  end

  # 3. Graph
  def test_graph_get
    g = @client.graph.get
    refute_nil g
  end

  def test_graph_search
    r = @client.graph.search(query: "test")
    refute_nil r
  end

  # 4. Entity CRUD
  def test_entity_crud
    name = "RUBY_TEST_ENTITY_#{SecureRandom.hex(3).upcase}"

    # Create
    created = @client.entities.create(
      entity_name: name,
      entity_type: "TEST",
      description: "Created by Ruby E2E",
      source_id: "ruby-e2e"
    )
    refute_nil created["status"]

    # List
    list = @client.entities.list
    refute_nil list["items"]

    # Get
    fetched = @client.entities.get(name: name)
    refute_nil fetched

    # Delete
    del = @client.entities.delete(name: name)
    refute_nil del["status"]
  end

  # 5. Relationships
  def test_relationships_list
    list = @client.relationships.list
    refute_nil list["items"]
  end

  # 6. Query
  def test_query
    r = @client.query.execute(query: "What is a knowledge graph?", mode: "hybrid")
    refute_nil r["answer"]
  end

  # 7. Chat
  def test_chat
    r = @client.chat.completions(message: "What entities exist?")
    refute_nil r["content"]
  rescue EdgeQuake::ApiError => e
    # Chat may require auth
    assert [401, 403].include?(e.status_code), "Unexpected error: #{e.message}"
  end

  # 8. Tenants
  def test_tenants_list
    list = @client.tenants.list
    refute_nil list["items"]
  end

  # 9. Users
  def test_users_list
    list = @client.users.list
    refute_nil list["users"]
  end

  # 10. API Keys
  def test_api_keys_list
    list = @client.api_keys.list
    refute_nil list["keys"]
  end

  # 11. Tasks
  def test_tasks_list
    list = @client.tasks.list
    refute_nil list["tasks"]
  end

  # 12. Pipeline Status
  def test_pipeline_status
    st = @client.pipeline.status
    refute_nil st.key?("is_busy")
  end

  # 13. Queue Metrics
  def test_queue_metrics
    m = @client.pipeline.queue_metrics
    refute_nil m["pending_count"]
    refute_nil m["active_workers"]
  end

  # 14. Models Catalog
  def test_models_catalog
    cat = @client.models.catalog
    refute_nil cat["providers"]
  end

  # 15. Models Health
  def test_models_health
    items = @client.models.health
    assert items.is_a?(Array), "models health should be an array"
    refute items.empty?, "models health should not be empty"
  end

  # 16. Provider Status
  def test_provider_status
    ps = @client.models.provider_status
    refute_nil ps["provider"]
  end

  # 17. Conversations
  def test_conversations_list
    list = @client.conversations.list
    refute_nil list
  end

  def test_conversations_create
    conv = @client.conversations.create(title: "Ruby E2E Test #{SecureRandom.hex(4)}")
    refute_nil conv
    assert(conv.key?("id") || conv.key?("conversation_id"), "should return conversation id")
  end

  # 18. Folders
  def test_folders_list
    list = @client.folders.list
    assert list.is_a?(Array), "folders should be an array"
  end

  def test_folders_create
    folder = @client.folders.create(name: "Ruby E2E Folder #{SecureRandom.hex(4)}")
    refute_nil folder
    assert(folder.key?("id") || folder.key?("name"), "should return folder")
  end

  # 19. Costs
  def test_costs_summary
    c = @client.costs.summary
    refute_nil c
  end

  # 20. Full Workflow
  def test_full_workflow
    # Upload
    doc = @client.documents.upload_text(
      title: "Ruby Workflow #{SecureRandom.hex(4)}",
      content: "Knowledge graphs connect entities through relationships."
    )
    refute_nil doc["document_id"]

    # Query
    qr = @client.query.execute(query: "What do knowledge graphs connect?", mode: "hybrid")
    refute_nil qr["answer"]

    # Pipeline status
    ps = @client.pipeline.status
    refute_nil ps.key?("is_busy")
  end
end

File: edgequake/sdks/swift/Tests/EdgeQuakeSDKTests/E2ETest.swift
import XCTest

@testable import EdgeQuakeSDK

/// End-to-end tests for EdgeQuakeSDK.
/// Requires a running EdgeQuake backend at localhost:8080.
/// Set EDGEQUAKE_BASE_URL to override.
final class E2ETest: XCTestCase {

    var client: EdgeQuakeClient!

    override func setUp() async throws {
        let base =
            ProcessInfo.processInfo.environment["EDGEQUAKE_BASE_URL"] ?? "http://localhost:8080"
        /// WHY: Default tenant/user IDs from database migration — avoids XCTSkip.
        let tenantId =
            ProcessInfo.processInfo.environment["EDGEQUAKE_TENANT_ID"]
                ?? "00000000-0000-0000-0000-000000000002"
        let userId =
            ProcessInfo.processInfo.environment["EDGEQUAKE_USER_ID"]
                ?? "00000000-0000-0000-0000-000000000001"
        let config = EdgeQuakeConfig(
            baseUrl: base,
            tenantId: tenantId,
            userId: userId
        )
        client = EdgeQuakeClient(config: config)
    }

    // MARK: - 1. Health

    func testHealthCheck() async throws {
        let h = try await client.health.check()
        XCTAssertEqual(h.status, "healthy")
        XCTAssertNotNil(h.version)
    }

    // MARK: - 2. Documents

    func testDocumentsListAndUpload() async throws {
        // List
        let list = try await client.documents.list()
        XCTAssertNotNil(list.documents)
        XCTAssertNotNil(list.total)

        // Upload
        let resp = try await client.documents.uploadText(
            title: "Swift SDK Test \(UUID().uuidString.prefix(8))",
            content: "Swift SDK integration test content. Knowledge graphs are powerful."
        )
        XCTAssertNotNil(resp.documentId)
        XCTAssertNotNil(resp.status)
    }

    // MARK: - 3. Graph

    func testGraphGet() async throws {
        let g = try await client.graph.get()
        XCTAssertNotNil(g)
    }

    func testGraphSearch() async throws {
        let r = try await client.graph.search(query: "test")
        XCTAssertNotNil(r)
    }

    // MARK: - 4. Entities CRUD

    func testEntityCrud() async throws {
        let entityName = "SWIFT_TEST_ENTITY_\(UUID().uuidString.prefix(6))"
        let req = CreateEntityRequest(
            entityName: entityName,
            entityType: "TEST",
            description: "Created by Swift E2E",
            sourceId: "swift-e2e"
        )
        let created = try await client.entities.create(req)
        XCTAssertNotNil(created.status)

        // List
        let list = try await client.entities.list()
        XCTAssertNotNil(list.items)

        // Get by name
        let fetched = try await client.entities.get(name: entityName)
        XCTAssertNotNil(fetched)

        // Delete
        let del = try await client.entities.delete(name: entityName)
        XCTAssertNotNil(del.status)
    }

    // MARK: - 5. Relationships

    func testRelationshipsList() async throws {
        let list = try await client.relationships.list()
        XCTAssertNotNil(list.items)
    }

    // MARK: - 6. Query

    func testQuery() async throws {
        let r = try await client.query.execute(
            query: "What is a knowledge graph?",
            mode: "hybrid"
        )
        XCTAssertNotNil(r.answer)
    }

    // MARK: - 7. Chat

    func testChat() async throws {
        do {
            let req = ChatCompletionRequest(message: "What entities exist?")
            let r = try await client.chat.completions(req)
            XCTAssertNotNil(r.content)
        } catch let e as EdgeQuakeError where e.statusCode == 401 || e.statusCode == 403 {
            // Chat may require auth — acceptable
            print("Chat: \(e.statusCode ?? 0) (expected if auth not configured)")
        }
    }

    // MARK: - 8. Tenants

    func testTenantsList() async throws {
        let list = try await client.tenants.list()
        XCTAssertNotNil(list.items)
    }

    // MARK: - 9. Users

    func testUsersList() async throws {
        let list = try await client.users.list()
        XCTAssertNotNil(list.users)
    }

    // MARK: - 10. API Keys

    func testApiKeysList() async throws {
        let list = try await client.apiKeys.list()
        XCTAssertNotNil(list.keys)
    }

    // MARK: - 11. Tasks

    func testTasksList() async throws {
        let list = try await client.tasks.list()
        XCTAssertNotNil(list.tasks)
    }

    // MARK: - 12. Pipeline Status

    func testPipelineStatus() async throws {
        let st = try await client.pipeline.status()
        XCTAssertNotNil(st.isBusy)
    }

    // MARK: - 13. Pipeline / Queue Metrics

    func testQueueMetrics() async throws {
        let m = try await client.pipeline.queueMetrics()
        XCTAssertNotNil(m.pendingCount)
        XCTAssertNotNil(m.activeWorkers)
    }

    // MARK: - 14. Models Catalog

    func testModelsCatalog() async throws {
        let catalog = try await client.models.catalog()
        XCTAssertNotNil(catalog.providers, "models catalog providers should not be nil")
    }

    // MARK: - 15. Models Health

    func testModelsHealth() async throws {
        let items = try await client.models.health()
        XCTAssertFalse(items.isEmpty, "models health should not be empty")
    }

    // MARK: - 16. Provider Status

    func testProviderStatus() async throws {
        let ps = try await client.models.providerStatus()
        XCTAssertNotNil(ps.provider)
    }

    // MARK: - 17. Conversations CRUD

    func testConversationsCRUD() async throws {
        // Create
        let conv = try await client.conversations.create(title: "Swift E2E Test \(UUID().uuidString.prefix(8))")
        XCTAssertNotNil(conv.id)
        print("Created conversation: \(conv.id ?? "nil") title=\(conv.title ?? "nil")")

        // List
        let convos = try await client.conversations.list()
        XCTAssertFalse(convos.isEmpty)
        print("Conversations: \(convos.count)")

        // Get detail
        let detail = try await client.conversations.get(id: conv.id!)
        XCTAssertNotNil(detail.conversation)
        XCTAssertEqual(detail.id, conv.id)

        // Delete (204 No Content)
        try await client.conversations.delete(id: conv.id!)
    }

    // MARK: - 18. Folders CRUD

    func testFoldersCRUD() async throws {
        // Create
        let folder = try await client.folders.create(name: "Swift E2E Folder \(UUID().uuidString.prefix(8))")
        XCTAssertNotNil(folder.id)
        print("Created folder: \(folder.id ?? "nil") name=\(folder.name ?? "nil")")

        // List
        let folders = try await client.folders.list()
        XCTAssertFalse(folders.isEmpty)
        print("Folders: \(folders.count)")

        // Delete (204 No Content)
        try await client.folders.delete(id: folder.id!)
    }

    // MARK: - 19. Costs

    func testCostsSummary() async throws {
        let c = try await client.costs.summary()
        XCTAssertNotNil(c)
    }

    // MARK: - 20. Full Workflow

    func testFullWorkflow() async throws {
        // Upload document
        let doc = try await client.documents.uploadText(
            title: "Swift Workflow \(UUID().uuidString.prefix(8))",
            content: "Knowledge graphs connect entities through relationships for better retrieval."
        )
        XCTAssertNotNil(doc.documentId)

        // Query
        let qr = try await client.query.execute(
            query: "What do knowledge graphs connect?",
            mode: "hybrid"
        )
        XCTAssertNotNil(qr.answer)

        // Verify pipeline status
        let ps = try await client.pipeline.status()
        XCTAssertNotNil(ps.isBusy)
    }
}
